planner/core/rule_inject_extra_projection.go

```diff
@@ -125,3 +131,70 @@ func injectProjBelowAgg(aggPlan PhysicalPlan, aggFuncs []*aggregation.AggFuncDes
 	aggPlan.SetChildren(proj)
 	return aggPlan
 }
+
+// The schema of PhysicalSort and PhysicalTopN are the same as the schema of
+// their children. When a projection is injected as the child of
+// PhysicalSort and PhysicalTopN, some extra columns will be added into the
+// schema of the Projection, thus we need to add another Projection upon
+// them to prune the redundant columns.
+func injectProjBelowSort(p PhysicalPlan, orderByItems []*ByItems) PhysicalPlan {
+	hasScalarFunc, numOrderByItems := false, len(orderByItems)
+	for i := 0; !hasScalarFunc && i < numOrderByItems; i++ {
+		_, isScalarFunc := orderByItems[i].Expr.(*expression.ScalarFunction)
+		hasScalarFunc = hasScalarFunc || isScalarFunc
+	}
+	if !hasScalarFunc {
+		return p
+	}
+
+	childPlan := p.Children()[0]
+	if childProj, isChildProj := childPlan.(*PhysicalProjection); isChildProj {
+		childProj.AvoidColumnEvaluator = true
+	}
+	projExprs := make([]expression.Expression, 0, p.Schema().Len())
+	for i := range p.Schema().Columns {
+		col := p.Schema().Columns[i]
+		col.Index = i
```

Comment:
Why do we need to set `Index` here? `ResolveIndices` has already done this work?

Reply:
We need, `ResolveIndices` was done in PhysicalOptimize, but injectExtraProjection is in PostOptimize.

Reply:
Hmm, I am still confused. Could you please give an example where `col.Index` would be different with `i` when coming here?

Reply:
@eurekaka 
1. We set `NewProj.Exprs` as the `originSortChild.Schema.Columns` to avoid the copy of the columns.
2. The `Column.Index` is not initialized for Schema. Thus we set the `Index` manually.